lib/teiserver/protocols/spring/spring_in.ex

```diff
@@ -1359,10 +1359,33 @@ defmodule Teiserver.Protocols.SpringIn do
     state
   end
 
-  defp do_handle("RING", username, _msg_id, state) do
-    userid = CacheUser.get_userid(username)
-    CacheUser.ring(userid, state.userid)
-    state
+  # https://springrts.com/dl/LobbyProtocol/ProtocolDescription.html#RING:client
+  # extended to accept originator as a second argument (only allowed when sent by bots)
+  # this allows spads to inform clients who originally rang them with spads-command !ring
+  defp do_handle("RING", data, _msg_id, state) do
+    case Regex.run(~r/(\S+) (\S+)/, data) do
+      [_, _, originator] ->
+        client = Client.get_client_by_id(state.userid)
+
+        cond do
+          client == nil ->
+            {:failure, "No client"}
+
+          not CacheUser.is_bot?(state.userid) ->
+            {:failure, "Not a bot"}
```

Comment:
`do_handle` must return the new state. If you return this tuple the caller's state is basically toast.
Since there's no simple mechanism to propagate this error I suggest you just ignore it.
So effectively, you can transform the `cond` into an `if`:

```elixir
if client != nil && not CacheUser.is_bot?(state.userid) do
...
end
```

Reply:
Thanks for the review.
I just copied this part from OPENBATTLE above:
https://github.com/beyond-all-reason/teiserver/blob/7cf54198dd2fa13f6ea415860c5870730972458e/lib/teiserver/protocols/spring/spring_in.ex#L810-L815
Does this need to be changed as well ?

Reply:
ooooh, I see.
That bit is fine because it's not returning this data directly. It is followed by [a check](https://github.com/beyond-all-reason/teiserver/blob/7cf54198dd2fa13f6ea415860c5870730972458e/lib/teiserver/protocols/spring/spring_in.ex#L851) and if there's a failure, it [replies with the correct error].

So if there's such a reply for the RING command, you should reply with that one.

Reply:
Should be fixed.